[#8419] Utilize configuration when populating gRPC options objects

ZacBlanco opened the pull request:
I've also removed the ttl and ttlAction fields from the setAttributes option as it is already included in the common options (#8417). Proto files have been regenerated.

Methods which need options passed on the client side should now utilize the `FileSystemOptions` class to generate the proper options based off of a configuration object. This will respect users' configuration on the client side. Otherwise options created with getDefaultInstance() don't respect a user's configuration with write type, ttl, etc.

Previously before this change the master would merge options with the client based on the configuration once an RPC is received. The new behavior populate all options on the client side, with any user-specified defaults overriding the configuration. All of the OperationContexts on the master side have had their constructors updated so that each now has a `defaults()` `create` and `mergeFrom`. `create` is to be used when accepting options from client requests `mergeFrom` and `defaults` should be used when a new context is needed on the master which doesn't come from the client options.

Fixes #8417 
Fixes #8419

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2075/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2076/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2077/
Test FAILed.

calvinjia commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -3585,14 +3585,17 @@ private void setAttributeSingleFile(RpcContext rpcContext, LockedInodePath inode
+        && protoOptions.getCommonOptions().hasTtlAction()) {
+      long ttl = protoOptions.getCommonOptions().getTtl();
+      if (inode.getTtl() != ttl
+          || inode.getTtlAction() != protoOptions.getCommonOptions().getTtlAction()) {
```
Can these conditions be cleaned up, it seems like we check `inode.getTtl()` twice, if `entry.setTtl` is idempotent, we could just set it regardless of whether or not it has changed (since the main thing is whether or not to update last mod time).

ZacBlanco replied:
While minor, the tradeoff is whether or not checking `inode.getTtl` twice is worth the tradeoff of setting the field on the entry. It looks like under the covers the entry does some other iother internal actions when you set a field. I don't know what the overhead is like, but doing that boolean comparison could possibly save some operations rather than setting it every time. An alternative so we don't do the comparison twice could be 

```
      long ttl = protoOptions.getCommonOptions().getTtl();
      boolean changedTtl = inode.getTtl() != ttl;
      if (changedTtl
          || inode.getTtlAction() != protoOptions.getCommonOptions().getTtlAction()) {
        if (changedTtl) {
          entry.setTtl(ttl);
        }
        entry.setLastModificationTimeMs(opTimeMs);
        entry.setTtlAction(ProtobufUtils.toProtobuf(
            protoOptions.getCommonOptions().getTtlAction()));
      }
```

calvinjia replied:
My suggestion is for improving code readability and not performance. I don't think the cost of either checking the boolean or setting the fields should be significant.

ZacBlanco commented on core/client/fs/src/main/java/alluxio/util/GrpcDefaultOptions.java:
```diff
@@ -0,0 +1,248 @@
+    return CreateDirectoryPOptions.newBuilder()
+        .setCommonOptions(getFileSystemMasterCommonPOptions(conf))
+        .setWriteType(conf.getEnum(PropertyKey.USER_FILE_WRITE_TYPE_DEFAULT, WritePType.class))
+        // No default configuration values, will pull from gRPC defaultInstance
```
I think it would be better to control non-configuration defaults from within the proto file rather than from these functions. Users don't necessarily have to go through this class to create these options and enforcing defaults from the proto definitions are a sure way to make sure that we always get the same defaults. Thoughts?

calvinjia replied:
We should make `getDefaultInstance` unavailable to end users. I think its better to control the defaults in one place rather than have some default in code and others in the proto definition. Also the proto def can be less flexible since it cannot be updated without recompilation.

ZacBlanco replied:
The issue I believe is that `getDefaultInstance()` isn't a function that we control. It's generated by protobuf. Obviously we could go in and manually change it to private but that means any time the file is regenerated we'd need to make that change. As far as I'm aware after glancing over the docs we can't control the availability of the `getDefaultInstance` function.

calvinjia replied:
We don't want to modify the autogenerated code. One soft method is to make `GrpcDefaultOptions` the public facing API. Users should only use the utils class when they need to get a default option, and if they decide to use the proto object directly they are on their own.

ZacBlanco replied:
Sorry I don't quite understand your suggestion. Are you saying that the client methods should take a different type instead (some kind of wrapper around the options)? or just that users should always use the `GrpcDefaultOptions` class?

calvinjia replied:
Users should always use the `GrpcDefaultOptions` class.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2082/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2083/
Test FAILed.

apc999 commented on core/base/src/main/java/alluxio/wire/LoadMetadataType.java:
```diff
@@ -24,15 +24,15 @@
    * Never loads metadata.
    */
-  Never(0),
+  NEVER(0),
```
good catch!

apc999 commented on core/client/fs/src/main/java/alluxio/util/GrpcDefaultOptions.java:
```diff
@@ -0,0 +1,259 @@
+   */
+  public static CreateDirectoryPOptions getCreateDirectoryPOptions(AlluxioConfiguration conf) {
+    return CreateDirectoryPOptions.newBuilder()
+        .setCommonOptions(getFileSystemMasterCommonPOptions(conf))
```
nit: keep these setters in alphabetical order so in the future when adding more fields, there will be unique locations to insert

apc999 commented on core/client/fs/src/main/java/alluxio/util/GrpcDefaultOptions.java:
```diff
@@ -0,0 +1,259 @@
+ * This class contains static methods which can be passed Alluxio configuration objects that
+ * will populate the gRPC options objects with the proper values based on the given configuration.
+ */
+public class GrpcDefaultOptions {
```
should this become a util class then with private constructors since it has no instance members.

ZacBlanco replied:
good catch

gpang commented:
@ZacBlanco is `Fixes 8415` the correct number? 8415 seems to point to another PR.

ZacBlanco commented:
God catch! It's supposed to be #8417

ggezer commented on core/client/fs/src/main/java/alluxio/util/GrpcDefaultOptions.java:
```diff
@@ -0,0 +1,259 @@
+ * This class contains static methods which can be passed Alluxio configuration objects that
+ * will populate the gRPC options objects with the proper values based on the given configuration.
+ */
+public class GrpcDefaultOptions {
```
We have a master counterpart of this class named FileSystemMasterOptions and it's a util class.
For the sake of consistency can we rename it as FileSystemClientOptions? We had this intermediary class before and it was named as such. This logic is not gRPC related so we don't need to name it as such.

ZacBlanco replied:
I can agree with that. However I don't necessarily see the logic here as client-side only issues. After taking a look at `FileSystemMasterOptions` we have much of the same code. It's just that the class relies on `ServerConfiguration` which is unavailable to the client.

I see this functionality as something that's shared on both on the client and server side. The logic between both classes is effectively the same. Originally I was going to put `GrpcDefaultOptions` in `core-common` but it relied on the `core-transport` dependency which I didn't want to add, so I moved it into the client. If we do want to add that dependency and move the logic to `core-common` then both client and server can rely on the same logic to populate the options objects. The server side will just need to pass `ServerConfiguration.global()` to all of the methods.

ggezer replied:
My motivation is to restrict this change's impact to client side. Server should have the authority to enforce defaults, and client should only send when it wants to override those defaults.

ZacBlanco commented:
Thanks for the comments @ggezer!

The issue surrounding these changes is that the Alluxio client isn't sending requests with the values configured via cluster defaults, site-properties, runtime, etc.. Calling `getDefaultInstance()` on any of the `POptions` classes simply returns the auto-generated gRPC defaults from our proto files. We should make sure that all requests which are sent, unless explicity modified, contain the options which respect a user's Alluxio configuration. Otherwise if the options aren't populated with the proper values we'll be sending requests that may not match what the user is expecting to be sending.

To address 1:

- I modified some tests too because the rationale is that pretty much anywhere in the codebase that uses `getDefaultInstance` should be populating values based on configuration, not based on the gRPC defaults. Thus, I changed tests where `getDefaultInstance` was utilized to make sure clients were sending the proper options. If there's a specific test in question you can point me to it.

To address 2:

- I don't know if I fully understand what you mean by "patching to BaseFileSystem" but I don't see how it would allow us to be more picky about which configuration sources to use. You're correct in that we can query sources of configuration, however I don't think the source of configuration applies in this situation. For example, if a client is sending a request, and the default TTL action is configured as `FREE` but the gRPC POptions default is `DELETE` the client should always override the POption default of `DELETE` regardless of the configuration source, no?

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2099/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2100/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2102/
Test FAILed.

calvinjia commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -436,7 +436,7 @@ private static MountInfo getRootMountInfo(MasterUfsManager ufsManager) {
           ServerConfiguration.getNestedProperties(PropertyKey.MASTER_MOUNT_TABLE_ROOT_OPTION);
       MountPOptions mountOptions = MountContext
-          .defaults(MountPOptions.newBuilder().setShared(shared).putAllProperties(rootUfsConf))
+          .mergeFrom(MountPOptions.newBuilder().setShared(shared).putAllProperties(rootUfsConf))
```
(Applies in general)
Should the `mergeFrom` be the other way around? This way will override the options from the mount context which should be respected?

ZacBlanco replied:
I'm not sure. Currently this will respect the previous implementation's behavior.

calvinjia replied:
Ah I see, the implemented `mergeFrom` is the opposite of the Protobuf `mergeFrom`. This is slightly confusing, could you leave a TODO to change the method name or have some other way of disambiguating this?

ggezer replied:
It's actually similar behaviour to protobuf, but I agree that it doesn't communicate on with what it's merging from. If we decide to change the name, all Context.mergeFrom() should be being used from master side, so it should be an easy refactor.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2106/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2108/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2109/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2110/
Test PASSed.

ggezer commented on core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java:
```diff
@@ -147,13 +148,15 @@ public boolean isClosed() {
   public void createDirectory(AlluxioURI path)
       throws FileAlreadyExistsException, InvalidPathException, IOException, AlluxioException {
-    createDirectory(path, CreateDirectoryPOptions.getDefaultInstance());
+    createDirectory(path, FileSystemOptions.createDirectoryDefaults(mFsContext.getConf()));
```
You don't need to pass the defaults since the options that you pass from here will be merged once again with the defaults. getDefaultInstance() should be fine in these utility versions.

ggezer replied:
Same for all utility methods (the ones without options) in BaseFileSystem.

ZacBlanco replied:
Good catch, fixed

ggezer commented on core/client/fs/src/main/java/alluxio/client/file/RetryHandlingFileSystemMasterClient.java:
```diff
@@ -232,7 +233,7 @@ public void mount(final AlluxioURI alluxioPath, final AlluxioURI ufsPath,
   @Override
   public void rename(final AlluxioURI src, final AlluxioURI dst)
```
Does this one even have any callers? I don't see a reason to have an utility at client level. Could you double check and remove if this method is redundant?

ZacBlanco replied:
Removing would require changing the public-facing API. I don't think removing this is in-scope for the PR.

ggezer replied:
This is not public API.

ZacBlanco replied:
Yes, but the public API calls this rename method

ggezer commented on core/client/fs/src/main/java/alluxio/util/FileSystemOptions.java:
```diff
@@ -0,0 +1,258 @@
+   * @param conf Alluxio configuration
+   * @return options based on the configuration
+   */
+  public static CreateDirectoryPOptions createDirectoryDefaults(AlluxioConfiguration conf) {
```
Did you consider returning builder instead? That will decrease the merging overhead at BaseFileSystem. 
Or it could be both. createDirectoryDefaults() vs createDirectoryDefaultsBuilder(), the former can call the latter.

ZacBlanco replied:
I considered it, but it would require another big change because all of the methods that call this will now need to use `build()` in a lot of places. I think the change would be for the better but I put it off because of the amount of work required.

If it's something you'd like to see I can implement it once I address the rest of the comments

ggezer commented on core/server/master/src/main/java/alluxio/master/file/FileSystemMasterOptions.java:
```diff
@@ -12,199 +12,24 @@
   /**
    * @return Master side defaults for {@link CompleteFilePOptions}
    */
   public static CompleteFilePOptions completeFileDefaults() {
```
Is this the only method that keeps this file alive? Can't this be moved to FileSystemOptions?

ZacBlanco replied:
It could be, but I also wanted to separate on the master v. client-side options because there are many options objects but only a few are used on client. Even with one method I felt it warranted enough difference to keep this file

ggezer commented on core/server/proxy/src/main/java/alluxio/proxy/PathsRestServiceHandler.java:
```diff
@@ -110,7 +111,7 @@ public Response createDirectory(@PathParam("path") final String path,
       if (options == null) {
         mFileSystem.createDirectory(new AlluxioURI(path),
-                CreateDirectoryPOptions.getDefaultInstance());
+                FileSystemOptions.createDirectoryDefaults(ServerConfiguration.global()));
```
This goes to FileSystem. Why do you need to supply the defaults here?

ggezer replied:
Same for all.

ZacBlanco replied:
Good catch. I changed this before we did the merging. Fixed

ggezer commented on job/server/src/main/java/alluxio/job/migrate/MigrateDefinition.java:
```diff
@@ -246,7 +246,7 @@ private static String computeTargetPath(String path, String source, String desti
     String newDir = computeTargetPath(path, source, destination);
     mFileSystem.createDirectory(new AlluxioURI(newDir),
-        CreateDirectoryPOptions.getDefaultInstance());
+        FileSystemOptions.createDirectoryDefaults(ServerConfiguration.global()));
```
Doesn't mFileSystem do the merging already?

ZacBlanco replied:
fixed

ggezer commented on shell/src/main/java/alluxio/cli/fs/command/CatCommand.java:
```diff
@@ -54,7 +55,7 @@ protected void runPlainPath(AlluxioURI path, CommandLine cl)
       throw new FileDoesNotExistException(ExceptionMessage.PATH_MUST_BE_FILE.getMessage(path));
     }
-    OpenFilePOptions options = OpenFilePOptions.getDefaultInstance();
+    OpenFilePOptions options = FileSystemOptions.openFileDefaults(mFileSystem.getConf());
```
This is passed to a FileSystem. Merging should be done there.

ZacBlanco replied:
fixed

ggezer commented on shell/src/main/java/alluxio/cli/fs/command/CheckConsistencyCommand.java:
```diff
@@ -109,7 +110,8 @@ public int run(CommandLine cl) throws AlluxioException, IOException {
   private void runConsistencyCheck(AlluxioURI path, boolean repairConsistency) throws
       AlluxioException, IOException {
     List<AlluxioURI> inconsistentUris =
-        checkConsistency(path, CheckConsistencyPOptions.getDefaultInstance());
```
FileSystem should do merging.

ZacBlanco replied:
The checkConsistency function uses the underlying client, not the public client API. So it still needs to merge options.

ggezer commented on shell/src/main/java/alluxio/cli/fs/command/CpCommand.java:
```diff
@@ -521,7 +522,8 @@ private void copy(AlluxioURI srcPath, AlluxioURI dstPath, boolean recursive)
   private void copyFile(AlluxioURI srcPath, AlluxioURI dstPath)
       throws AlluxioException, IOException {
     try (Closer closer = Closer.create()) {
-      OpenFilePOptions openFileOptions = OpenFilePOptions.getDefaultInstance();
```
FileSystem should do merging.

ZacBlanco replied:
fixed

ggezer commented on shell/src/main/java/alluxio/cli/fs/command/CpCommand.java:
```diff
@@ -738,7 +740,7 @@ private void copyFileToLocal(AlluxioURI srcPath, AlluxioURI dstPath)
     File tmpDst = new File(outputFile.getPath() + randomSuffix);
 
     try (Closer closer = Closer.create()) {
-      OpenFilePOptions options = OpenFilePOptions.getDefaultInstance();
```
FileSystem should do merging.

ZacBlanco replied:
fixed

ggezer commented on shell/src/main/java/alluxio/cli/fs/command/FileSystemCommandUtils.java:
```diff
@@ -43,7 +44,9 @@ private FileSystemCommandUtils() {} // prevent instantiation
   public static void setTtl(FileSystem fs, AlluxioURI path, long ttlMs,
       TtlAction ttlAction) throws AlluxioException, IOException {
     SetAttributePOptions options = SetAttributePOptions.newBuilder().setRecursive(true)
-        .setTtl(ttlMs).setTtlAction(ttlAction).build();
```
FileSystem should do merging.

ZacBlanco replied:
The user is passing a specific ttl+action here we want to override options

ggezer commented on shell/src/main/java/alluxio/cli/fs/command/HeadCommand.java:
```diff
@@ -70,7 +71,7 @@ protected void runPlainPath(AlluxioURI plainPath, CommandLine cl)
     if (status.isFolder()) {
       throw new IOException(ExceptionMessage.PATH_MUST_BE_FILE.getMessage(plainPath));
     }
-    OpenFilePOptions options = OpenFilePOptions.getDefaultInstance();
```
FileSystem should do merging.

ZacBlanco replied:
fixed

ggezer commented on shell/src/main/java/alluxio/cli/fs/command/TailCommand.java:
```diff
@@ -68,7 +69,7 @@ protected void runPlainPath(AlluxioURI path, CommandLine cl)
     if (status.isFolder()) {
       throw new IOException(ExceptionMessage.PATH_MUST_BE_FILE.getMessage(path));
     }
-    OpenFilePOptions options = OpenFilePOptions.getDefaultInstance();
```
FileSystem should do merging.

ZacBlanco replied:
fixed

ggezer commented on tests/src/main/java/alluxio/master/backcompat/ops/PersistFile.java:
```diff
@@ -37,10 +38,10 @@ public void apply(Clients clients) throws Exception {
     FileSystem fs = clients.getFileSystemClient();
     Utils.createFile(fs, PATH);
     clients.getFileSystemMasterClient().scheduleAsyncPersist(PATH,
-        ScheduleAsyncPersistencePOptions.getDefaultInstance());
```
FileSystem should do merging.

ZacBlanco replied:
Not here, we don't use a `FileSystem` object. Options must be specified to this client.

ggezer replied:
Right. You're calling the client. My bad.
I was consumed by Copy/Paste rage...

ZacBlanco replied:
It happens to the best of us

ggezer commented on tests/src/test/java/alluxio/client/cli/fs/FileSystemShellUtilsTest.java:
```diff
@@ -125,9 +126,11 @@ public static String resetFileHierarchy(FileSystem fs, WritePType writeType)
-    fs.createDirectory(new AlluxioURI(TEST_DIR));
-    fs.createDirectory(new AlluxioURI(TEST_DIR + "/foo"));
-    fs.createDirectory(new AlluxioURI(TEST_DIR + "/bar"));
+    CreateDirectoryPOptions dirOptions = CreateDirectoryPOptions.getDefaultInstance().toBuilder()
```
Why did you have to change this now?

ZacBlanco replied:
Because the client configuration write type is set as `CACHE_THROUGH` and when the directories are created we expect them to be in a not-persisted state. Without passing the write type here the directories weren't being created with the proper write type and the assertion was failing for tests which checked that the directory wasn't persisted after creating the file hierarchy

ggezer commented on tests/src/test/java/alluxio/client/cli/fs/command/LsCommandIntegrationTest.java:
```diff
@@ -393,10 +394,10 @@ public void lsWithExtendedAcl() throws IOException, AlluxioException {
 
     mFileSystem.setAcl(new AlluxioURI("/testRoot/testDir"), SetAclAction.MODIFY,
         Arrays.asList(AclEntry.fromCliString("default:user:nameduser:rwx")),
-        SetAclPOptions.getDefaultInstance());
```
FileSystem should do merging.

ZacBlanco replied:
fixed

ggezer commented on tests/src/test/java/alluxio/client/fs/BufferedBlockInStreamIntegrationTest.java:
```diff
@@ -62,7 +64,8 @@ public static void beforeClass() throws Exception {
 
   private static List<CreateFilePOptions> getOptionSet() {
     List<CreateFilePOptions> ret = new ArrayList<>(3);
-    CreateFilePOptions optionsDefault = CreateFilePOptions.getDefaultInstance();
```
FileSystem should do merging.

ZacBlanco replied:
fixed

ggezer commented on tests/src/test/java/alluxio/client/fs/ConcurrentFileSystemMasterSetTtlTest.java:
```diff
@@ -133,7 +134,9 @@ public void run() {
             AuthenticatedClientUser.set(TEST_USER);
             barrier.await();
             mFileSystem.setAttribute(paths[iteration], SetAttributePOptions.newBuilder()
-                .setTtl(ttls[iteration]).setTtlAction(TtlAction.DELETE).build());
```
FileSystem should do merging.

ZacBlanco replied:
We want to specifically set the TTL and Ttl action here which varies from iteration to iteration. Passing options here is required. It's only changed because setAttribute no longer has TTL and TTL action as a top-level attribute. They both exists in common options now

ggezer commented on tests/src/test/java/alluxio/client/fs/FileSystemIntegrationTest.java:
```diff
@@ -334,8 +335,8 @@ public void createExistingDirectory() throws Exception {
     AlluxioURI path = new AlluxioURI("/dir");
-    FileSystemTestUtils.createByteFile(mFileSystem, path, CreateFilePOptions.getDefaultInstance(),
-        10);
+    FileSystemTestUtils.createByteFile(mFileSystem, path,
```
FileSystem should do merging.

ZacBlanco replied:
fixed

ggezer commented on tests/src/test/java/alluxio/client/fs/LoadMetadataIntegrationTest.java:
```diff
@@ -174,35 +174,33 @@ public void loadMetadataOnceAfterUfsDelete() throws Exception {
 
   @Test
   public void loadAlwaysConfiguration() throws Exception {
```
Whole test goes through FileSystem. So merging should supply the defaults.

ZacBlanco replied:
The issue is that the clients are created after the cluster is started and before we have a chance to set configuration, so these tests need to specifically set the options in order for their values to be respected

ZacBlanco replied:
However after closer inspection I think I changed how this test expected to work.

I've restored the original functionality by adding cluster resource config annotations to those tests

```
  @LocalAlluxioClusterResource.Config(
      confParams = {
          PropertyKey.Name.USER_FILE_METADATA_LOAD_TYPE, "NEVER/ONCE/ALWAYS"
      })
```

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2116/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2117/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2120/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2121/
Test FAILed.

ZacBlanco commented:
jenkins, test this please

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2122/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2124/
Test PASSed.
